R/filter-nm-data.R

```diff
@@ -0,0 +1,193 @@
+  expr <- gsub("(?<=[^=!<>])=(?=[^=!<>])", "==", expr, perl = TRUE)
+  return(expr)
+}
+
+#' Function to invert ignore expressions
+#' @noRd
+#' @keywords internal
+invert_operator <- function(expr) {
+  expr <- dplyr::case_when(
+    grepl("==", expr) ~ gsub("==", "!=", expr),
+    grepl("!=", expr) ~ gsub("!=", "==", expr),
+    grepl("<=", expr) ~ gsub("<=", ">=", expr),
+    grepl(">=", expr) ~ gsub(">=", "<=", expr),
+    grepl("<", expr) ~ gsub("<", ">", expr),
+    grepl(">", expr) ~ gsub(">", "<", expr),
+    TRUE ~ expr
+  )
+  return(expr)
+}
+
+#' Translate `IGNORE` and `ACCEPT` filter expressions into [dplyr::filter()]
+#'  expressions.
+#'
+#' @param nm_expr a `NONMEM` filter expression. e.g., `'ID.EQ.2, BLQ=1'`.
+#' @param type Either `'ignore'` or `'accept'`. Denotes which type of `NONMEM`
+#'  filtering the expression corresponds to.
+#' @param data_cols column names associated with the input data. Used for
+#'  `'ignore'` expressions.
+#'
+#' @examples
+#' \dontrun{
+#' test_exprs <- "SEX==1, ID.EQ.2, WT/=70, AGE.NE.30, A=1, WT.GT.40"
+#'
+#' translate_nm_expr(test_exprs, type = 'ignore')
+#' #> [1] "SEX!=1 & ID!=2 & WT==70 & AGE==30 & A!=1 & WT<40"
+#'
+#' translate_nm_expr(test_exprs, type = 'accept')
+#' #> [1] "SEX==1 & ID==2 & WT!=70 & AGE!=30 & A==1 & WT>40"
+#'
+#'
+#' # Use of `@`, `#`, or form `IGNORE=C2` require `data_cols` to be specified,
+#' # though only the first column is used
+#' data_cols <- c("C", "ID", "TIME", "EVID", "DV", "BLQ")
+#'
+#' translate_nm_expr("#", data_cols = data_cols)
+#' #> [1] "!grepl('^#', C)"
+#'
+#' translate_nm_expr("c2", data_cols = data_cols)
+#' #> [1] "C!='c2'"
+#'
+#' translate_nm_expr("@", data_cols = data_cols)
+#' #> [1] "!grepl('^[A-Za-z@]', C)"
+#' }
+#' @keywords internal
+translate_nm_expr <- function(
+    nm_expr,
+    type = c("ignore", "accept"),
+    data_cols = NULL
+){
+  type <- match.arg(type)
+  checkmate::assert_character(data_cols, min.len = 1, null.ok = TRUE)
+
+  # Translate NM operators and separate any (list) type expressions
+  exprs <- translate_nm_operator(nm_expr)
+  exprs <- stringr::str_split(exprs, ",")[[1]] %>% stringr::str_trim()
+
+  r_exprs <- purrr::map_chr(exprs, function(expr){
+    if(type == "ignore"){
+      # `IGNORE=#`, `IGNORE=@`, `IGNORE=c1`, `IGNORE=(list)`
+      if(expr == "#"){
+        # IGNORE=# is the default.  That is, in the absence of IGNORE option, any
+        # record whose first character is '#' is treated as a comment record.
+        paste0("!grepl('^#', ", data_cols[1], ")")
+      }else if(expr == "@"){
+        # IGNORE=@ signifies that any data record having an alphabetic character
+        # or `@` as its first non-blank character in column one should be ignored.
+        #  - This permits a table file having header lines to be used as an
+        #    NM-TRAN data set.
+        paste0("!grepl('^[A-Za-z@]', ", data_cols[1], ")")
+      }else if(grepl('^[a-zA-Z0-9]{1,}$', expr)){
```

Comment:
I dont think so, _I think_ this can be any number of characters as long as it's _at least 1_:
> IGNORE=c1 [Note: The following two options, IGNORE  and  ACCEPT  allow  the user  to drop records from the NM-TRAN data set prior to the run. With NONMEM 7.5, it is also possible to  drop  records  from  the NONMEM  data  set  using abbreviated code, which is more flexible than the criteria that can be specified in the $DATA record. See PRED_IGNORE_DATA, below.] 

> Specifies that any data record having the character c1 in  column 1  should be ignored, i.e., these records are not included in the NONMEM data set.  This allows comment records to be  included  in the  NM-TRAN  data set.  In general, records having the character c1 in column 1 will be called "comment records". Also permitted: IGNORE='c' or IGNORE="c",  where  c  may  be  any character  except  space.   IGNORE=# is the default.  That is, in the absence of IGNORE option, any record whose first character is # is treated as a comment record.

Will do some testing to confirm this

Reply:
It looks like it should be a single character: 

```
^[a-zA-Z]$
```

Reply:
Sounds good. Yeah the digits definitely shouldnt have been in there either